core/bin/dust_api.rs

```diff
@@ -2702,6 +2703,170 @@ async fn tables_rows_list(
     }
 }
 
+#[derive(serde::Deserialize)]
+struct FoldersUpsertPayload {
+    folder_id: String,
+    timestamp: Option<u64>,
+    parents: Vec<String>,
+    title: String,
+}
+
+async fn folders_upsert(
+    Path((project_id, data_source_id)): Path<(i64, String)>,
+    State(state): State<Arc<APIState>>,
+    Json(payload): Json<FoldersUpsertPayload>,
+) -> (StatusCode, Json<APIResponse>) {
+    let project = project::Project::new_from_id(project_id);
+
+    let folder = Folder {
+        data_source_id: data_source_id.to_string(),
+        folder_id: payload.folder_id.clone(),
+        created: utils::now(),
+    };
+
+    match state
+        .store
+        .upsert_data_source_folder(&project, &data_source_id, &folder)
```

Comment:
I think we'll want to unify at some point in store.rs. Some methods take objects in input (`Folder`, `DataSource` , `Node`.. ) but for example table takes the values directly.

IMO the latter is much more convenient:
- you don't have to pass `created`, which is always "now"
- if you are in a `match` or an `if`, you cannot create a temporary `Folder` and pass it by reference to the method because it gets dropped so the ref outlives the actual object (see the "maybe_node" variable I had to create in table upsert)

(not saying we should fix here ofc)

Reply:
The current standard is objects though no?

Consistency matters here I think.

Reply:
Consistency matters 100%. I think we should change to params for all, for the reasons mentioned above (mostly ref passing of dropped value). Alternative is to take an owned value but this also has downsides.

There's not many changes to make and I'm happy to take if you are ok with it.

Reply:
You mean creating an object inline does not work because of the ref? And that's the annoying part?

The owned value could make sense but it would create weirdness in data source source code I think with some weird unwanted clones...

Exploding the params might be a bit verbose at places no?

Reply:
The maybe_node is a bit weirdish I agree but 100% solved once we get rid of the Option.

Agree that this will happen again though

Reply:
Yeah that specific thing goes away, but having a setup that makes it hard to do things conditionally and forces us to systematically pass to upserts/inserts all properties we get back from gets/lists is not ideal IMHO

Reply:
Owned value is fine if it only contains refs right?

Reply:
Yep but then it has to be called DocumentProps and we're back at writing typescript :D

Reply:
More seriously might be a good tradeoff and we could have helper functions such as Table.props() ?

But fuck that stink typescript